toolkits/google/arcade_google/tools/calendar.py

```diff
@@ -356,3 +361,140 @@ async def delete_event(
         f"{notification_message}"
     )
+
+
+# TODO: would be nice to have a "min_slot_duration" parameter
+# TODO: find a way to have "include_weekends" parameter without confusing LLMs
+@tool(
+    requires_auth=Google(
+        scopes=["https://www.googleapis.com/auth/calendar.readonly"],
+    ),
+)
+async def find_time_slots_when_everyone_is_free(
+    context: ToolContext,
+    email_addresses: Annotated[
+        Optional[list[str]],
+        "The list of email addresses from people in the same organization domain (apart from the "
+        "currently logged in user) to search for free time slots. Defaults to None, which will "
+        "return free time slots for the current user only.",
+    ] = None,
+    start_date: Annotated[
+        Optional[str],
+        "The start date to search for time slots in the format 'YYYY-MM-DD'. Defaults to today's "
+        "date. It will search starting from this date at the time 00:00:00.",
+    ] = None,
+    end_date: Annotated[
+        Optional[str],
+        "The end date to search for time slots in the format 'YYYY-MM-DD'. Defaults to seven days "
+        "from the start date. It will search until this date at the time 23:59:59.",
+    ] = None,
+    start_time_boundary: Annotated[
+        str,
+        "Will return free slots in any given day starting from this time in the format 'HH:MM'. "
+        "Defaults to '08:00', which is a usual business hour start time.",
+    ] = "08:00",
+    end_time_boundary: Annotated[
+        str,
+        "Will return free slots in any given day until this time in the format 'HH:MM'. "
+        "Defaults to '18:00', which is a usual business hour end time.",
+    ] = "18:00",
+) -> Annotated[
+    dict,
+    "A dictionary with the free slots and the timezone in which time slots are represented.",
+]:
+    """
+    Provides time slots when everyone is free within a given date range and time boundaries.
+    """
+    credentials = Credentials(
+        context.authorization.token if context.authorization and context.authorization.token else ""
+    )
+
+    # Build google api services
+    oauth_service = build("oauth2", "v2", credentials=credentials)
+    calendar_service = build("calendar", "v3", credentials=credentials)
+
+    email_addresses = email_addresses or []
+
+    if isinstance(email_addresses, str):
+        email_addresses = [email_addresses]
+
+    # Add the currently logged in user to the list of email addresses
+    user_info = oauth_service.userinfo().get().execute()
+    if user_info["email"] not in email_addresses:
+        email_addresses.append(user_info["email"])
+
+    # Get the timezone of the currently logged in user
+    calendar = calendar_service.calendars().get(calendarId="primary").execute()
+    timezone_name = calendar.get("timeZone")
+
+    try:
+        tz = ZoneInfo(timezone_name)
+    # If the calendar timezone name is not supported by Python's zoneinfo, use UTC
```

Comment:
which calendar timezone names are not supported? If we had this list then couldn't we write the `timezone_name` to `tz` conversion for those unsupported timezones? i.e., we would never need to default to UTC?

Reply:
In principle, both Google Calendar and Python zoneinfo module implement the [IANA tz database](https://www.iana.org/time-zones) and this exception should never be raised.

But we don't control Google responses (and I learned to not trust what 3pty APIs say in their docs). And we don't control the system's tz database implementation (where the engine is running). This is usually in `/usr/share/zoneinfo` or `/usr/share/zoneinfo.default`.

We can simply let an exception be raised when the two tz database implementations (Google & system) have any incompatibility. Or we can default to UTC and carry on the user request...